File: src/shared/attributes.ts
// Helpers da árvore de atributos por herói (H12; dado puro em attributeData.ts).
// Monta o layout por herói (colunas = grupos), formata o efeito de cada nó e cruza com os
// níveis alocados lidos do save (attributeSaveDatas → snapshot.heroAttributes).
import { ATTR_GROUPS, ATTR_NODES, type AttrNode } from './attributeData'
import type { HeroAttributeLevel } from './types'

function fmt(n: number): string {
  return new Intl.NumberFormat('pt-BR', { maximumFractionDigits: 2 }).format(n)
}

// Rótulos pt-BR dos tipos de habilidade ativa (act do datamine); fallback no próprio act.
const ACT_LABELS: Record<string, string> = {
  BASEATTACK: 'Ataque básico',
  BASEATTACK_COUNT: 'Ataque básico',
  COOLDOWN: 'Habilidade ativa'
}

export function actLabel(act: string | null | undefined): string {
  if (!act) return 'Habilidade ativa'
  return ACT_LABELS[act] ?? 'Habilidade ativa'
}

/** Rótulo do nó (passivo: nome do status; ativo: nome curto da habilidade). */
export function attrTitle(node: AttrNode): string {
  if (node.kind === 'active') return actLabel(node.act)
  return node.name ?? node.st ?? `Nó ${node.id}`
}

/** Linha-template do efeito com um valor preenchido (passivos). */
function fillLine(line: string | null | undefined, value: number, fallbackName: string): string {
  const tmpl = line ?? `${fallbackName} +{0}`
  return tmpl.includes('{0}') ? tmpl.replace('{0}', fmt(value)) : `${tmpl} ${fmt(value)}`
}

/** Efeito por ponto (passivo) ou faixa de escala (ativo), legível em pt-BR. */
export function attrPerLevel(node: AttrNode): string {
  if (node.kind === 'active') {
    const dmg = node.dmg ?? []
    if (dmg.length === 0) return '—'
    const first = dmg[0]
    const last = dmg[dmg.length - 1]
    return first === last ? `valor ${fmt(first)}` : `valor ${fmt(first)} → ${fmt(last)} por nível`
  }
  return fillLine(node.line, node.v ?? 0, attrTitle(node)) + ' por ponto'
}

/** Efeito acumulado no nível alocado (passivo: v×nível; ativo: valor do nível). */
export function attrEffectAtLevel(node: AttrNode, level: number): string {
  if (level <= 0) return attrPerLevel(node)
  if (node.kind === 'active') {
    const dmg = node.dmg ?? []
    const idx = Math.min(level, dmg.length) - 1
    const meta = [node.delivery, node.dmgType].filter(Boolean).join(' · ')
    const cd = node.cd != null ? ` · recarga ${node.cd}s` : ''
    return idx >= 0 ? `valor ${fmt(dmg[idx])}${meta ? ` (${meta})` : ''}${cd}` : meta || '—'
  }
  return fillLine(node.line, (node.v ?? 0) * level, attrTitle(node))
}

// --- Layout por herói ---

export interface AttrNodeView {
  node: AttrNode
  level: number // nível alocado (0 = não alocado)
  effect: string // efeito no nível atual (ou por ponto se 0)
  perLevel: string // efeito por ponto / escala
}

export interface AttrColumn {
  grp: number
  x: number
  nodes: AttrNodeView[]
}

export interface HeroAttrTree {
  hero: number
  columns: AttrColumn[]
  totalAllocated: number // soma dos níveis alocados
  allocatedNodes: number // nós com nível > 0
  totalNodes: number
}

/** Heróis presentes no catálogo de atributos (ordenados). */
export function attrHeroes(): number[] {
  return [...new Set(ATTR_NODES.map((n) => n.hero))].sort((a, b) => a - b)
}

/** Nós de um herói (ordenados por grupo e id). */
export function nodesForHero(hero: number): AttrNode[] {
  return ATTR_NODES.filter((n) => n.hero === hero).sort(
    (a, b) => a.gx - b.gx || a.id - b.id
  )
}

/**
 * Monta a árvore de atributos de um herói em colunas (uma por grupo), cruzando com os níveis
 * alocados (do save). `allocated` mapeia id do nó → nível.
 */
export function heroAttributeTree(
  hero: number,
  allocated: HeroAttributeLevel[]
): HeroAttrTree {
  const levelByKey = new Map<number, number>()
  for (const a of allocated) levelByKey.set(a.key, a.level)

  const colMap = new Map<number, AttrColumn>()
  let totalAllocated = 0
  let allocatedNodes = 0

  for (const node of nodesForHero(hero)) {
    const level = levelByKey.get(node.id) ?? 0
    if (level > 0) {
      totalAllocated += level
      allocatedNodes++
    }
    let col = colMap.get(node.grp)
    if (!col) {
      col = { grp: node.grp, x: node.gx, nodes: [] }
      colMap.set(node.grp, col)
    }
    col.nodes.push({
      node,
      level,
      effect: attrEffectAtLevel(node, level),
      perLevel: attrPerLevel(node)
    })
  }

  const columns = [...colMap.values()].sort((a, b) => a.x - b.x)
  const totalNodes = columns.reduce((sum, c) => sum + c.nodes.length, 0)
  return { hero, columns, totalAllocated, allocatedNodes, totalNodes }
}

export { ATTR_GROUPS, ATTR_NODES }
export type { AttrNode }

File: dotnet/TbhTracker.Core/Logic/Attributes.cs
using System.Globalization;
using TbhTracker.Core.Models;

namespace TbhTracker.Core.Logic;

// Port de src/shared/attributes.ts (catalogo cru vem de attributeData.json).

public sealed class AttrNodeView
{
    public AttrNode Node { get; init; } = new();
    public int Level { get; init; }
    public string Effect { get; init; } = "";
    public string PerLevel { get; init; } = "";
}

public sealed class AttrColumn
{
    public int Grp { get; init; }
    public int X { get; init; }
    public List<AttrNodeView> Nodes { get; init; } = new();
}

public sealed class HeroAttrTree
{
    public int Hero { get; init; }
    public List<AttrColumn> Columns { get; init; } = new();
    public int TotalAllocated { get; init; }
    public int AllocatedNodes { get; init; }
    public int TotalNodes { get; init; }
}

public static class Attributes
{
    private static readonly CultureInfo PtBr = CultureInfo.GetCultureInfo("pt-BR");

    private static string Fmt(double n) => n.ToString("#,##0.##", PtBr);

    private static readonly Dictionary<string, string> ActLabels = new()
    {
        ["BASEATTACK"] = "Ataque básico",
        ["BASEATTACK_COUNT"] = "Ataque básico",
        ["COOLDOWN"] = "Habilidade ativa"
    };

    public static string ActLabel(string? act)
    {
        if (string.IsNullOrEmpty(act)) return "Habilidade ativa";
        return ActLabels.TryGetValue(act, out var l) ? l : "Habilidade ativa";
    }

    public static string AttrTitle(AttrNode node)
    {
        if (node.Kind == "active") return ActLabel(node.Act);
        return node.Name ?? node.St ?? $"Nó {node.Id}";
    }

    private static string FillLine(string? line, double value, string fallbackName)
    {
        var tmpl = line ?? $"{fallbackName} +{{0}}";
        return tmpl.Contains("{0}") ? tmpl.Replace("{0}", Fmt(value)) : $"{tmpl} {Fmt(value)}";
    }

    public static string AttrPerLevel(AttrNode node)
    {
        if (node.Kind == "active")
        {
            var dmg = node.Dmg ?? new List<double>();
            if (dmg.Count == 0) return "—";
            var first = dmg[0];
            var last = dmg[^1];
            return first == last ? $"valor {Fmt(first)}" : $"valor {Fmt(first)} → {Fmt(last)} por nível";
        }
        return FillLine(node.Line, node.V ?? 0, AttrTitle(node)) + " por ponto";
    }

    public static string AttrEffectAtLevel(AttrNode node, int level)
    {
        if (level <= 0) return AttrPerLevel(node);
        if (node.Kind == "active")
        {
            var dmg = node.Dmg ?? new List<double>();
            var idx = Math.Min(level, dmg.Count) - 1;
            var meta = string.Join(" · ", new[] { node.Delivery, node.DmgType }.Where(x => !string.IsNullOrEmpty(x)));
            var cd = node.Cd != null ? $" · recarga {Fmt(node.Cd.Value)}s" : "";
            if (idx >= 0)
                return $"valor {Fmt(dmg[idx])}{(meta.Length > 0 ? $" ({meta})" : "")}{cd}";
            return meta.Length > 0 ? meta : "—";
        }
        return FillLine(node.Line, (node.V ?? 0) * level, AttrTitle(node));
    }

    public static List<int> AttrHeroes() =>
        Catalog.Attributes.Nodes.Select(n => n.Hero).Distinct().OrderBy(h => h).ToList();

    public static List<AttrNode> NodesForHero(int hero) =>
        Catalog.Attributes.Nodes
            .Where(n => n.Hero == hero)
            .OrderBy(n => n.Gx).ThenBy(n => n.Id)
            .ToList();

    public static HeroAttrTree HeroAttributeTree(int hero, IEnumerable<HeroAttributeLevel> allocated)
    {
        var levelByKey = new Dictionary<int, int>();
        foreach (var a in allocated) levelByKey[a.Key] = a.Level;

        var colMap = new Dictionary<int, (int Grp, int X, List<AttrNodeView> Nodes)>();
        var colOrder = new List<int>();
        var totalAllocated = 0;
        var allocatedNodes = 0;

        foreach (var node in NodesForHero(hero))
        {
            var level = levelByKey.TryGetValue(node.Id, out var lv) ? lv : 0;
            if (level > 0)
            {
                totalAllocated += level;
                allocatedNodes++;
            }
            if (!colMap.TryGetValue(node.Grp, out var col))
            {
                col = (node.Grp, node.Gx, new List<AttrNodeView>());
                colMap[node.Grp] = col;
                colOrder.Add(node.Grp);
            }
            col.Nodes.Add(new AttrNodeView
            {
                Node = node,
                Level = level,
                Effect = AttrEffectAtLevel(node, level),
                PerLevel = AttrPerLevel(node)
            });
        }

        var columns = colMap.Values
            .OrderBy(c => c.X)
            .Select(c => new AttrColumn { Grp = c.Grp, X = c.X, Nodes = c.Nodes })
            .ToList();
        var totalNodes = columns.Sum(c => c.Nodes.Count);

        return new HeroAttrTree
        {
            Hero = hero,
            Columns = columns,
            TotalAllocated = totalAllocated,
            AllocatedNodes = allocatedNodes,
            TotalNodes = totalNodes
        };
    }
}
